### PR TITLE
Support for additional labels and annotations to be inherited by tenant services

### LINKED ISSUE #79: Support for additional labels and annotations to be inherited by tenant services
# Describe the feature
The idea is to leave an additional section in the Tenant manifest where the Cluster admin can specify additional labels and/or annotations to apply to the services created in the tenant. Initially, there are no specific semantics assigned to these labels, they just will be assigned to the services and endpoints in the tenant. This can be useful to the adopters of Capsule to implement their specific use cases.

```yaml
apiVersion: capsule.clastix.io/v1alpha1
kind: Tenant
metadata:
  name: oil
spec:
  additionalServiceLabels:
    capsule.clastix.io/do_stuff: "always"
```
...

# What would the new user story look like?

1. The cluster admin creates a tenant with this additional fields
2. Tenant owner creates a new _Namespace_
3. Tenant owner creates new services in the tenant namespace
3. All the services and related endpoints created in this tenant will inherit the labels
4. Depending on the case, additional business logics can be implemented by using the labels. 

Please note the specific business logic is not part of the Capsule. Instead it is left to the cluster admin and or to the tenant owner.

# Expected behavior
All the services and related endpoints created in the tenant inherit the labels.

### PR DESCRIPTION
**Pull request by MaxFedotov**
Closes #79 

@prometherion  This realization is purely on a webhook. I tried to move update events to controller, but in order to achieve it, we had to create 3 controllers for each of the resources types (svc\ep\eps). So If it is necessary, I think we had to discuss realization one more time :)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -103,7 +103,7 @@ ifeq (, $(shell which kustomize))
 	KUSTOMIZE_GEN_TMP_DIR=$$(mktemp -d) ;\
 	cd $$KUSTOMIZE_GEN_TMP_DIR ;\
 	go mod init tmp ;\
-	go get sigs.k8s.io/kustomize/kustomize/v3@v3.5.4 ;\
+	go get sigs.k8s.io/kustomize/kustomize/v3@v3.8.1 ;\
```
Comment (prometherion) [inline Makefile:106]: Can we document why we need to pin another version? Just for the sake of history :)

File: api/v1alpha1/tenant_types.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ import (
 // +kubebuilder:validation:Minimum=1
 type NamespaceQuota uint
 
-type NamespaceMetadata struct {
+type AddtionalMetadata struct {
```
Comment (prometherion) [inline api/v1alpha1/tenant_types.go:28]: Typo

File: pkg/webhook/service_labels/patching.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+func (h *handler) syncLabels(ctx context.Context, req admission.Request, client client.Client, decoder *admission.Decoder) admission.Response {
+	var availableLabels, availableAnnotations map[string]string
+	var namespaceName string
+	var patch []jsonpatch.JsonPatchOperation
+
+	switch req.Kind.Kind {
+	case "Service":
+		svc := &corev1.Service{}
```
Comment (prometherion) [inline pkg/webhook/service_labels/patching.go:88]: I don't want to be picky but I guess the most elegant solution would be a type embedding in order to decorate these types and void all of these duplicate lines of codes [as I did here](https://github.com/clastix/capsule/blob/a4b0c0fe27906621af61093e05964f9a676f0c01/pkg/webhook/ingress/types.go#L17-L71)

Tl;dr; embedding Service, Endpoints and EndpointSlice into a domain struct adhering to an Interface providing some funcs as `GetNamespace`, `GetAnnotations`, `GetLabels`.

File: pkg/webhook/service_labels/patching.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+*/
+
+package service_labels
+
+import (
+	"context"
+	"net/http"
+	"strings"
+
+	capsulev1alpha1 "github.com/clastix/capsule/api/v1alpha1"
```
Comment (prometherion) [inline pkg/webhook/service_labels/patching.go:24]: Please, `goimports` here

**Review by prometherion — CHANGES_REQUESTED** [review]
Some little changes but LGTM once fixed: let's plan for the future to add a Controller since, right now, would be an overkill.

Last but not least: nice catch with JSON Pointer! 🥇

File: Makefile (hunk 1 above)
Comment (prometherion) [reply Makefile:106]: Ah, e2e is not working here due to this update :(

```
namespace/system created
customresourcedefinition.apiextensions.k8s.io/tenants.capsule.clastix.io created
clusterrole.rbac.authorization.k8s.io/capsule-proxy-role created
clusterrole.rbac.authorization.k8s.io/capsule-metrics-reader created
clusterrolebinding.rbac.authorization.k8s.io/capsule-manager-rolebinding created
clusterrolebinding.rbac.authorization.k8s.io/capsule-proxy-rolebinding created
mutatingwebhookconfiguration.admissionregistration.k8s.io/capsule-mutating-webhook-configuration created
validatingwebhookconfiguration.admissionregistration.k8s.io/capsule-validating-webhook-configuration created
Error from server (NotFound): error when creating "STDIN": namespaces "capsule-system" not found
Error from server (NotFound): error when creating "STDIN": namespaces "capsule-system" not found
Error from server (NotFound): error when creating "STDIN": namespaces "capsule-system" not found
Error from server (NotFound): error when creating "STDIN": namespaces "capsule-system" not found
Error from server (NotFound): error when creating "STDIN": namespaces "capsule-system" not found
```

Comment (MaxFedotov) [reply Makefile:106]: seems like kustomize wasn't removed
```
sudo snap remove kustomize
  sudo snap remove kustomize
  shell: /bin/bash -e {0}
snap "kustomize" is not installed
```

and correct version of kustomize wasn't downloaded
```
kind load docker-image --nodes capsule-control-plane --name capsule quay.io/clastix/capsule:latest
Image: "quay.io/clastix/capsule:latest" with ID "sha256:7598fa7cfb473d786eb8c1d573308a7a5c4d2ac7b6e6a8632d94feeada3014d0" not present on node "capsule-control-plane"
make deploy
make[1]: Entering directory '/home/runner/work/capsule/capsule'
/home/runner/go/bin/controller-gen "crd:trivialVersions=true" rbac:roleName=manager-role webhook paths="./..." output:crd:artifacts:config=config/crd/bases
cd config/manager && /usr/local/bin/kustomize edit set image controller=quay.io/clastix/capsule:latest
/usr/local/bin/kustomize build config/default | kubectl apply -f -
namespace/system created
```

As there is no 
```
go: creating new go.mod: module tmp
go: downloading sigs.k8s.io/kustomize v3.8.1
```

And this version is needed because we need this PR https://github.com/kubernetes-sigs/kustomize/commit/74b45392fc1194e6003a1f1ab97dae96fb85a76d#diff-f058c1a370fe4d82b2864c52fe8e9554
In order to fix ordering when running `kustomize build`.

And version 3.8.2 had problems with https://github.com/operator-framework/operator-sdk/issues/3826

Comment (MaxFedotov) [reply Makefile:106]: But, maybe it was due to another issue (related to failurePolicy=Ignore), l will check it in a few hours :)

Comment (MaxFedotov) [reply Makefile:106]: Yep, this not needed, v3.5.4 is OK. Will rollback this change

File: pkg/webhook/service_labels/patching.go (hunk 4 above)
Comment (MaxFedotov) [reply pkg/webhook/service_labels/patching.go]: fixed

File: pkg/webhook/service_labels/patching.go (hunk 3 above)
Comment (MaxFedotov) [reply pkg/webhook/service_labels/patching.go]: thanks for the good advice, done :)

File: api/v1alpha1/tenant_types.go (hunk 2 above)
Comment (MaxFedotov) [reply api/v1alpha1/tenant_types.go]: fixed

File: Makefile (hunk 1 above)
Comment (MaxFedotov) [reply Makefile]: @prometherion 
I've rollbacked changes to Makefile, but there is still an error. Could you please take a look?

**Review by prometherion — APPROVED** [review]
LGTM, many kudos for landing this!
